{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This note is used for debugging Tony's perturb pipeline "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/oak/stanford/groups/engreitz/Users/shshanhe/tools/miniforge3/envs/torch-cNMF/bin/python\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import sys\n",
    "print(sys.executable)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# bug in read excel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pool</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>expected_cells</th>\n",
       "      <th>min_umi_threshold</th>\n",
       "      <th>fastq_dir</th>\n",
       "      <th>sample_to_well_mapping</th>\n",
       "      <th>i7_barcode</th>\n",
       "      <th>i5_barcode</th>\n",
       "      <th>notes</th>\n",
       "      <th>read1_length</th>\n",
       "      <th>read2_length</th>\n",
       "      <th>paired_guide_pool</th>\n",
       "      <th>paired_guide_sample</th>\n",
       "      <th>paired_guide_sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>eRZ58</td>\n",
       "      <td>eRZ58_GEX40B</td>\n",
       "      <td>eRZ58:eRZ58_GEX40B</td>\n",
       "      <td>gex</td>\n",
       "      <td>5000</td>\n",
       "      <td>50</td>\n",
       "      <td>/oak/stanford/groups/engreitz/Users/ronghao/Se...</td>\n",
       "      <td>plate1</td>\n",
       "      <td>CTTGTA</td>\n",
       "      <td>CTGGCA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>64</td>\n",
       "      <td>58</td>\n",
       "      <td>eRZ58</td>\n",
       "      <td>eRZ58_gRNA40B</td>\n",
       "      <td>eRZ58:eRZ58_gRNA40B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>eRZ58</td>\n",
       "      <td>eRZ58_gRNA40B</td>\n",
       "      <td>eRZ58:eRZ58_gRNA40B</td>\n",
       "      <td>guide</td>\n",
       "      <td>5000</td>\n",
       "      <td>50</td>\n",
       "      <td>/oak/stanford/groups/engreitz/Users/ronghao/Se...</td>\n",
       "      <td>plate1</td>\n",
       "      <td>TCGGCA</td>\n",
       "      <td>GGTAGC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>64</td>\n",
       "      <td>58</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    pool         sample            sample_id sample_type  expected_cells  \\\n",
       "0  eRZ58   eRZ58_GEX40B   eRZ58:eRZ58_GEX40B         gex            5000   \n",
       "1  eRZ58  eRZ58_gRNA40B  eRZ58:eRZ58_gRNA40B       guide            5000   \n",
       "\n",
       "   min_umi_threshold                                          fastq_dir  \\\n",
       "0                 50  /oak/stanford/groups/engreitz/Users/ronghao/Se...   \n",
       "1                 50  /oak/stanford/groups/engreitz/Users/ronghao/Se...   \n",
       "\n",
       "  sample_to_well_mapping i7_barcode i5_barcode  notes  read1_length  \\\n",
       "0                 plate1     CTTGTA     CTGGCA    NaN            64   \n",
       "1                 plate1     TCGGCA     GGTAGC    NaN            64   \n",
       "\n",
       "   read2_length paired_guide_pool paired_guide_sample paired_guide_sample_id  \n",
       "0            58             eRZ58       eRZ58_gRNA40B    eRZ58:eRZ58_gRNA40B  \n",
       "1            58               NaN                 NaN                    NaN  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_k562=pd.read_excel(\"references_yann_k562/sample_info.k562_yann.xlsx\")\n",
    "df_eRZ_20A=pd.read_excel(\"/oak/stanford/groups/engreitz/Users/ronghao/Sequencing/20251218_eRZ59_Split-seq_4M_Novaseq/perturb_pipeline/analysis/references_eRZ59/sample_info_eRZ58_20A.xlsx\")\n",
    "df_eRZ_40B=pd.read_excel(\"/oak/stanford/groups/engreitz/Users/ronghao/Sequencing/20251218_eRZ59_Split-seq_4M_Novaseq/perturb_pipeline/analysis/references_eRZ59/sample_info_eRZ58_40B.xlsx\")\n",
    "df_k562\n",
    "df_eRZ_20A\n",
    "df_eRZ_40B"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# bug in ribosomal anotation, need to be modified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting gffutils\n",
      "  Downloading gffutils-0.13-py3-none-any.whl.metadata (1.5 kB)\n",
      "Collecting pyfaidx>=0.5.5.2 (from gffutils)\n",
      "  Downloading pyfaidx-0.9.0.3-py3-none-any.whl.metadata (25 kB)\n",
      "Collecting argh>=0.26.2 (from gffutils)\n",
      "  Downloading argh-0.31.3-py3-none-any.whl.metadata (7.4 kB)\n",
      "Collecting argcomplete>=1.9.4 (from gffutils)\n",
      "  Downloading argcomplete-3.6.3-py3-none-any.whl.metadata (16 kB)\n",
      "Collecting simplejson (from gffutils)\n",
      "  Downloading simplejson-3.20.2-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.2 kB)\n",
      "Requirement already satisfied: packaging in /oak/stanford/groups/engreitz/Users/shshanhe/tools/miniforge3/envs/torch-cNMF/lib/python3.9/site-packages (from pyfaidx>=0.5.5.2->gffutils) (25.0)\n",
      "Downloading gffutils-0.13-py3-none-any.whl (1.6 MB)\n",
      "\u001b[2K   \u001b[90m‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m55.4 MB/s\u001b[0m  \u001b[33m0:00:00\u001b[0m\n",
      "\u001b[?25hDownloading argcomplete-3.6.3-py3-none-any.whl (43 kB)\n",
      "Downloading argh-0.31.3-py3-none-any.whl (44 kB)\n",
      "Downloading pyfaidx-0.9.0.3-py3-none-any.whl (29 kB)\n",
      "Downloading simplejson-3.20.2-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (137 kB)\n",
      "Installing collected packages: simplejson, pyfaidx, argh, argcomplete, gffutils\n",
      "\u001b[2K   \u001b[90m‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ\u001b[0m \u001b[32m5/5\u001b[0m [gffutils]4/5\u001b[0m [gffutils]te]\n",
      "\u001b[1A\u001b[2KSuccessfully installed argcomplete-3.6.3 argh-0.31.3 gffutils-0.13 pyfaidx-0.9.0.3 simplejson-3.20.2\n"
     ]
    }
   ],
   "source": [
    "#import gffutils\n",
    "from collections import Counter\n",
    "import random\n",
    "import sys\n",
    "!{sys.executable} -m pip install gffutils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: ENSMUSG00000102628\n",
      "seqid: 1 start: 150956201 end: 150958296 strand: +\n",
      "attribute keys: ['gene_biotype', 'gene_id', 'gene_name', 'gene_source', 'gene_version']\n",
      "gene_biotype: TEC\n",
      "gene_id: ENSMUSG00000102628\n",
      "gene_type: \n",
      "hgnc_id: \n",
      "gene_name: Gm37671\n"
     ]
    }
   ],
   "source": [
    "import gffutils\n",
    "\n",
    "gene_database_path = \"/oak/stanford/groups/engreitz/Users/shshanhe/project/exchange_perturb_project/data/reference/Mus_musculus.GRCm39.108.annotation.db\"\n",
    "db = gffutils.FeatureDB(gene_database_path)\n",
    "g = next(db.features_of_type(\"gene\"))\n",
    "print(\"ID:\", g.id)\n",
    "# ID: ENSMUSG00000102628\n",
    "print(\"seqid:\", g.seqid, \"start:\", g.start, \"end:\", g.end, \"strand:\", g.strand)\n",
    "# seqid: 1 start: 150956201 end: 150958296 strand: +\n",
    "print(\"attribute keys:\", sorted(g.attributes.keys()))\n",
    "gene_id = \"ENSMUSG00000102628\"\n",
    "g\n",
    "print(\"gene_biotype:\", g.attributes.get(\"gene_biotype\", [\"\"])[0])\n",
    "print(\"gene_id:\", g.attributes.get(\"gene_id\", [\"\"])[0])\n",
    "print(\"gene_type:\", g.attributes.get(\"gene_type\", [\"\"])[0])\n",
    "print(\"hgnc_id:\", g.attributes.get(\"hgnc_id\",  [\"\"])[0])\n",
    "print(\"gene_name:\", g.attributes.get(\"gene_name\",[\"\"])[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: ENSG00000290825.1\n",
      "seqid: chr1 start: 11869 end: 14409 strand: +\n",
      "attribute keys: ['gene_id', 'gene_name', 'gene_type', 'level', 'tag']\n",
      "gene_biotype: \n",
      "gene_id: ENSG00000290825.1\n",
      "gene_type: lncRNA\n",
      "hgnc_id: \n",
      "gene_name: DDX11L2\n"
     ]
    }
   ],
   "source": [
    "# human\n",
    "gene_database_path = \"/oak/stanford/groups/engreitz/Users/tonyzeng/GW_PERTURB/references/gencode.v46.annotation.db\"\n",
    "db = gffutils.FeatureDB(gene_database_path)\n",
    "g = next(db.features_of_type(\"gene\"))\n",
    "print(\"ID:\", g.id)\n",
    "print(\"seqid:\", g.seqid, \"start:\", g.start, \"end:\", g.end, \"strand:\", g.strand)\n",
    "print(\"attribute keys:\", sorted(g.attributes.keys()))\n",
    "\n",
    "gene_id = \"ENSG00000142541\"\n",
    "g\n",
    "print(\"gene_biotype:\", g.attributes.get(\"gene_biotype\", [\"\"])[0])\n",
    "print(\"gene_id:\", g.attributes.get(\"gene_id\", [\"\"])[0])\n",
    "print(\"gene_type:\", g.attributes.get(\"gene_type\", [\"\"])[0])\n",
    "print(\"hgnc_id:\", g.attributes.get(\"hgnc_id\",  [\"\"])[0])\n",
    "print(\"gene_name:\", g.attributes.get(\"gene_name\",[\"\"])[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_comprehensive_gene_annotations(adata, ribosomal_genes_path, gene_database_path, cell_cycle_genes_path):\n",
    "    \"\"\"Add comprehensive gene annotations to var dataframe.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    adata : AnnData\n",
    "        Annotated data object\n",
    "    ribosomal_genes_path : str\n",
    "        Path to ribosomal genes file\n",
    "    gene_database_path : str\n",
    "        Path to gene database file\n",
    "    cell_cycle_genes_path : str\n",
    "        Path to cell cycle genes file\n",
    "    \"\"\"\n",
    "    log_print(\"üß¨ Adding comprehensive gene annotations...\")\n",
    "\n",
    "    # Create reference files dict for internal use\n",
    "    REFERENCE_FILES = {\n",
    "        \"ribosomal_genes\": ribosomal_genes_path,\n",
    "        \"gene_database\": gene_database_path,\n",
    "        \"cell_cycle_genes\": cell_cycle_genes_path\n",
    "    }\n",
    "\n",
    "    # Load ribosomal genes from tab-separated file - use \"HGNC ID\" column\n",
    "    # ribosomal_hgnc_ids = set()\n",
    "    with open(REFERENCE_FILES[\"ribosomal_genes\"]) as f:\n",
    "        ribosomal_genes_list = [x.strip() for x in f]\n",
    "    ribosomal_genes = set(ribosomal_genes_list)  # For fast lookup\n",
    "\n",
    "        # header = f.readline().strip().split('\\t')\n",
    "        # hgnc_col = header.index(\"HGNC ID\")\n",
    "        # for line in f:\n",
    "        #     cols = line.strip().split('\\t')\n",
    "        #     if len(cols) > hgnc_col:\n",
    "        #         ribosomal_hgnc_ids.add(cols[hgnc_col])\n",
    "\n",
    "    log_print(f\"  üìã Loaded {len(ribosomal_genes)} ribosomal gene\")\n",
    "    log_print(f\"  üìã Sample ribosomal gene: {list(ribosomal_genes)[:5]}\")\n",
    "\n",
    "    # Load gencode database\n",
    "    db = gffutils.FeatureDB(REFERENCE_FILES[\"gene_database\"])\n",
    "\n",
    "    # Initialize annotation columns\n",
    "    adata.var['gene_type'] = ''\n",
    "    # adata.var['hgnc_id'] = ''\n",
    "    adata.var['gene'] = ''  # Gene symbols\n",
    "    adata.var['chromosome'] = ''\n",
    "    adata.var['ribosomal'] = False\n",
    "    adata.var['cell_cycle'] = False\n",
    "    adata.var['cell_cycle_phase'] = ''\n",
    "    adata.var['pct_cells_expressed'] = 0.0\n",
    "\n",
    "    # Calculate % cells expressed (using expression threshold > 0)\n",
    "    log_print(\"  üìä Calculating % cells expressed...\")\n",
    "    cells_expressed = (adata.X > 0).sum(axis=0)\n",
    "    if hasattr(cells_expressed, 'A1'):\n",
    "        cells_expressed = cells_expressed.A1\n",
    "    adata.var['pct_cells_expressed'] = (cells_expressed / adata.n_obs) * 100\n",
    "\n",
    "    # Load cell cycle genes for annotation\n",
    "    with open(REFERENCE_FILES[\"cell_cycle_genes\"]) as f:\n",
    "        cell_cycle_genes_list = [x.strip() for x in f]\n",
    "    cell_cycle_genes = set(cell_cycle_genes_list)  # For fast lookup\n",
    "    s_phase_genes = set(cell_cycle_genes_list[:43])  # First 43 are S phase\n",
    "\n",
    "    # Annotate from gencode database first\n",
    "    log_print(\"  üóÉÔ∏è  Loading gene annotations from gencode database...\")\n",
    "    log_print(f\"  üîç Sample gene IDs from adata.var.index: {list(adata.var.index[:5])}\")\n",
    "\n",
    "    annotated_count = 0\n",
    "    missing_count = 0\n",
    "    error_count = 0\n",
    "\n",
    "    for i, gene_id in enumerate(adata.var.index):\n",
    "        if i < 5:  # Debug first 5 genes\n",
    "            log_print(f\"  üîç Trying to annotate gene {i+1}: {gene_id}\")\n",
    "\n",
    "        try:\n",
    "            # Query database for this gene by ID\n",
    "            gene = db[gene_id]\n",
    "        except gffutils.exceptions.FeatureNotFoundError:\n",
    "            gene = db[gene_id.split(\".\", 1)[0]]\n",
    "            \n",
    "            if i < 5:\n",
    "                log_print(f\"    ‚úÖ Found gene {gene_id} in database\")\n",
    "\n",
    "            # Extract annotations - handle missing attributes gracefully\n",
    "            # NOTE: We use .get() here because not all genes have all attributes (e.g., some lack hgnc_id)\n",
    "            # This is expected and OK - we should NOT raise errors for missing optional attributes\n",
    "            gene_type = gene.attributes.get('gene_biotype', [''])[0]\n",
    "            #hgnc_id = gene.attributes.get('hgnc_id', [''])[0]\n",
    "            gene_name = gene.attributes.get('gene_name', [''])[0]\n",
    "\n",
    "            if i < 5:\n",
    "                log_print(f\"    üìã Gene info: type={gene_type}, name={gene_name}, chr={gene.seqid}\")\n",
    "\n",
    "            # Update annotations\n",
    "            adata.var.loc[gene_id, 'gene_type'] = gene_type\n",
    "            # adata.var.loc[gene_id, 'hgnc_id'] = hgnc_id\n",
    "            adata.var.loc[gene_id, 'gene'] = gene_name  # Add gene symbols\n",
    "            adata.var.loc[gene_id, 'chromosome'] = gene.seqid\n",
    "\n",
    "            # Check ribosomal genes (by HGNC ID)\n",
    "            if gene_name in ribosomal_genes:\n",
    "                adata.var.loc[gene_id, 'ribosomal'] = True\n",
    "\n",
    "            # Check cell cycle genes (by gene name)\n",
    "            if gene_name in cell_cycle_genes:\n",
    "                adata.var.loc[gene_id, 'cell_cycle'] = True\n",
    "                # Determine phase using pre-computed S phase set\n",
    "                if gene_name in s_phase_genes:\n",
    "                    adata.var.loc[gene_id, 'cell_cycle_phase'] = 'S'\n",
    "                else:\n",
    "                    adata.var.loc[gene_id, 'cell_cycle_phase'] = 'G2M'\n",
    "\n",
    "            annotated_count += 1\n",
    "\n",
    "        except Exception as e:\n",
    "            raise RuntimeError(f\"Failed to annotate gene {gene_id}: {e}\") from e\n",
    "\n",
    "    log_print(f\"  üìä Annotation results: {annotated_count} success, {missing_count} missing, {error_count} errors\")\n",
    "\n",
    "    # CRITICAL: Fail if gene annotation completely failed\n",
    "    if annotated_count == 0:\n",
    "        raise RuntimeError(f\"CRITICAL: Gene annotation completely failed. 0/{adata.n_vars} genes annotated. Database lookup is broken.\")\n",
    "\n",
    "    # CRITICAL: Fail if most genes failed annotation\n",
    "    success_rate = annotated_count / adata.n_vars\n",
    "    if success_rate < 0.5:  # Less than 50% success\n",
    "        raise RuntimeError(f\"CRITICAL: Gene annotation mostly failed. Only {success_rate:.1%} ({annotated_count}/{adata.n_vars}) genes annotated successfully.\")\n",
    "\n",
    "    # Calculate mitochondrial genes using annotated gene symbols\n",
    "    log_print(\"  üß¨ Calculating mitochondrial gene scores...\")\n",
    "    mt_genes = adata.var[\"gene\"].str.startswith((\"MT-\", \"mt-\"), na=False)\n",
    "    log_print(f\"  üìä Found {mt_genes.sum()} mitochondrial genes\")\n",
    "\n",
    "    if mt_genes.sum() > 0:\n",
    "        mt_gene_names = adata.var_names[mt_genes].tolist()\n",
    "\n",
    "        log_print(f\"  ‚úÖ Added MT percentage and score using {len(mt_gene_names)} genes\")\n",
    "    else:\n",
    "        raise RuntimeError(\"CRITICAL: No mitochondrial genes found for scoring. MT genes are required for quality control.\")\n",
    "\n",
    "    # Calculate ribosomal scores using annotated genes\n",
    "    log_print(\"  üß¨ Calculating ribosomal gene scores...\")\n",
    "    ribo_genes = adata.var['ribosomal'] == True\n",
    "    log_print(f\"  üìä Found {ribo_genes.sum()} ribosomal genes\")\n",
    "\n",
    "    if ribo_genes.sum() > 0:\n",
    "        ribo_gene_names = adata.var_names[ribo_genes].tolist()\n",
    "\n",
    "        # Calculate ribosomal percentage\n",
    "        adata.obs['pct_counts_ribo'] = (adata[:, ribo_genes].X.sum(axis=1) / adata.X.sum(axis=1)) * 100\n",
    "        if hasattr(adata.obs['pct_counts_ribo'], 'A1'):\n",
    "            adata.obs['pct_counts_ribo'] = adata.obs['pct_counts_ribo'].A1\n",
    "\n",
    "        log_print(f\"  ‚úÖ Added ribosomal percentage and score using {len(ribo_gene_names)} genes\")\n",
    "    else:\n",
    "        raise RuntimeError(\"CRITICAL: No ribosomal genes found for scoring. Ribosomal gene metrics are required for quality control.\")\n",
    "\n",
    "    # Summary statistics\n",
    "    log_print(f\"  ‚úÖ Annotated {annotated_count}/{adata.n_vars} genes\")\n",
    "    log_print(f\"  üìä Gene types: {adata.var['gene_type'].value_counts().head().to_dict()}\")\n",
    "    log_print(f\"  üìä Ribosomal: {adata.var['ribosomal'].sum()}\")\n",
    "    log_print(f\"  üìä Cell cycle: {adata.var['cell_cycle'].sum()}\")\n",
    "    log_print(f\"  üìä Chromosomes: {adata.var['chromosome'].nunique()} unique\")\n",
    "    log_print(f\"  üìä Mean % cells expressed: {adata.var['pct_cells_expressed'].mean():.1f}%\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_comprehensive_gene_annotations(adata, ribosomal_genes_path, gene_database_path, cell_cycle_genes_path):\n",
    "    \"\"\"Add comprehensive gene annotations to var dataframe.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    adata : AnnData\n",
    "        Annotated data object\n",
    "    ribosomal_genes_path : str\n",
    "        Path to ribosomal genes file\n",
    "    gene_database_path : str\n",
    "        Path to gene database file\n",
    "    cell_cycle_genes_path : str\n",
    "        Path to cell cycle genes file\n",
    "    \"\"\"\n",
    "    print(\"üß¨ Adding comprehensive gene annotations...\")\n",
    "\n",
    "    # Create reference files dict for internal use\n",
    "    REFERENCE_FILES = {\n",
    "        \"ribosomal_genes\": ribosomal_genes_path,\n",
    "        \"gene_database\": gene_database_path,\n",
    "        \"cell_cycle_genes\": cell_cycle_genes_path\n",
    "    }\n",
    "\n",
    "    # Load ribosomal genes from tab-separated file - use \"HGNC ID\" column\n",
    "    # ribosomal_hgnc_ids = set()\n",
    "    with open(REFERENCE_FILES[\"ribosomal_genes\"]) as f:\n",
    "        ribosomal_genes_list = [x.strip() for x in f]\n",
    "    ribosomal_genes = set(ribosomal_genes_list)  # For fast lookup\n",
    "\n",
    "        # header = f.readline().strip().split('\\t')\n",
    "        # hgnc_col = header.index(\"HGNC ID\")\n",
    "        # for line in f:\n",
    "        #     cols = line.strip().split('\\t')\n",
    "        #     if len(cols) > hgnc_col:\n",
    "        #         ribosomal_hgnc_ids.add(cols[hgnc_col])\n",
    "\n",
    "    print(f\"  üìã Loaded {len(ribosomal_genes)} ribosomal gene\")\n",
    "    print(f\"  üìã Sample ribosomal gene: {list(ribosomal_genes)[:5]}\")\n",
    "\n",
    "    # Load gencode database\n",
    "    db = gffutils.FeatureDB(REFERENCE_FILES[\"gene_database\"])\n",
    "\n",
    "    # Initialize annotation columns\n",
    "    adata.var['gene_type'] = ''\n",
    "    # adata.var['hgnc_id'] = ''\n",
    "    adata.var['gene'] = ''  # Gene symbols\n",
    "    adata.var['chromosome'] = ''\n",
    "    adata.var['ribosomal'] = False\n",
    "    adata.var['cell_cycle'] = False\n",
    "    adata.var['cell_cycle_phase'] = ''\n",
    "    adata.var['pct_cells_expressed'] = 0.0\n",
    "\n",
    "    # Calculate % cells expressed (using expression threshold > 0)\n",
    "    print(\"  üìä Calculating % cells expressed...\")\n",
    "    cells_expressed = (adata.X > 0).sum(axis=0)\n",
    "    if hasattr(cells_expressed, 'A1'):\n",
    "        cells_expressed = cells_expressed.A1\n",
    "    adata.var['pct_cells_expressed'] = (cells_expressed / adata.n_obs) * 100\n",
    "\n",
    "    # Load cell cycle genes for annotation\n",
    "    with open(REFERENCE_FILES[\"cell_cycle_genes\"]) as f:\n",
    "        cell_cycle_genes_list = [x.strip() for x in f]\n",
    "    cell_cycle_genes = set(cell_cycle_genes_list)  # For fast lookup\n",
    "    s_phase_genes = set(cell_cycle_genes_list[:43])  # First 43 are S phase\n",
    "\n",
    "    # Annotate from gencode database first\n",
    "    print(\"  üóÉÔ∏è  Loading gene annotations from gencode database...\")\n",
    "    print(f\"  üîç Sample gene IDs from adata.var.index: {list(adata.var.index[:5])}\")\n",
    "\n",
    "    annotated_count = 0\n",
    "    missing_count = 0\n",
    "    error_count = 0\n",
    "\n",
    "    for i, gene_id in enumerate(adata.var.index):\n",
    "        if i < 5:  # Debug first 5 genes\n",
    "            print(f\"  üîç Trying to annotate gene {i+1}: {gene_id}\")\n",
    "\n",
    "        try:\n",
    "            # Query database for this gene by ID\n",
    "            gene = db[gene_id]\n",
    "        except gffutils.exceptions.FeatureNotFoundError:\n",
    "            gene = db[gene_id.split(\".\", 1)[0]]\n",
    "            \n",
    "            if i < 5:\n",
    "            \n",
    "                print(f\"    ‚úÖ Found gene {gene_id} in database\")\n",
    "\n",
    "            # Extract annotations - handle missing attributes gracefully\n",
    "            # NOTE: We use .get() here because not all genes have all attributes (e.g., some lack hgnc_id)\n",
    "            # This is expected and OK - we should NOT raise errors for missing optional attributes\n",
    "            gene_type = gene.attributes.get('gene_biotype', [''])[0]\n",
    "            #hgnc_id = gene.attributes.get('hgnc_id', [''])[0]\n",
    "            gene_name = gene.attributes.get('gene_name', [''])[0]\n",
    "\n",
    "            if i < 5:\n",
    "                print(f\"    üìã Gene info: type={gene_type}, name={gene_name}, chr={gene.seqid}\")\n",
    "\n",
    "            # Update annotations\n",
    "            adata.var.loc[gene_id, 'gene_type'] = gene_type\n",
    "            # adata.var.loc[gene_id, 'hgnc_id'] = hgnc_id\n",
    "            adata.var.loc[gene_id, 'gene'] = gene_name  # Add gene symbols\n",
    "            adata.var.loc[gene_id, 'chromosome'] = gene.seqid\n",
    "\n",
    "            # Check ribosomal genes (by HGNC ID)\n",
    "            if gene_name in ribosomal_genes:\n",
    "                adata.var.loc[gene_id, 'ribosomal'] = True\n",
    "\n",
    "            # Check cell cycle genes (by gene name)\n",
    "            if gene_name in cell_cycle_genes:\n",
    "                adata.var.loc[gene_id, 'cell_cycle'] = True\n",
    "                # Determine phase using pre-computed S phase set\n",
    "                if gene_name in s_phase_genes:\n",
    "                    adata.var.loc[gene_id, 'cell_cycle_phase'] = 'S'\n",
    "                else:\n",
    "                    adata.var.loc[gene_id, 'cell_cycle_phase'] = 'G2M'\n",
    "\n",
    "            annotated_count += 1\n",
    "\n",
    "        except Exception as e:\n",
    "            raise RuntimeError(f\"Failed to annotate gene {gene_id}: {e}\") from e\n",
    "\n",
    "    print(f\"  üìä Annotation results: {annotated_count} success, {missing_count} missing, {error_count} errors\")\n",
    "\n",
    "    # CRITICAL: Fail if gene annotation completely failed\n",
    "    if annotated_count == 0:\n",
    "        raise RuntimeError(f\"CRITICAL: Gene annotation completely failed. 0/{adata.n_vars} genes annotated. Database lookup is broken.\")\n",
    "\n",
    "    # CRITICAL: Fail if most genes failed annotation\n",
    "    success_rate = annotated_count / adata.n_vars\n",
    "    if success_rate < 0.5:  # Less than 50% success\n",
    "        raise RuntimeError(f\"CRITICAL: Gene annotation mostly failed. Only {success_rate:.1%} ({annotated_count}/{adata.n_vars}) genes annotated successfully.\")\n",
    "\n",
    "    # Calculate mitochondrial genes using annotated gene symbols\n",
    "    print(\"  üß¨ Calculating mitochondrial gene scores...\")\n",
    "    mt_genes = adata.var[\"gene\"].str.startswith((\"MT-\", \"mt-\"), na=False)\n",
    "    print(f\"  üìä Found {mt_genes.sum()} mitochondrial genes\")\n",
    "\n",
    "    if mt_genes.sum() > 0:\n",
    "        mt_gene_names = adata.var_names[mt_genes].tolist()\n",
    "\n",
    "        print(f\"  ‚úÖ Added MT percentage and score using {len(mt_gene_names)} genes\")\n",
    "    else:\n",
    "        raise RuntimeError(\"CRITICAL: No mitochondrial genes found for scoring. MT genes are required for quality control.\")\n",
    "\n",
    "    # Calculate ribosomal scores using annotated genes\n",
    "    print(\"  üß¨ Calculating ribosomal gene scores...\")\n",
    "    ribo_genes = adata.var['ribosomal'] == True\n",
    "    print(f\"  üìä Found {ribo_genes.sum()} ribosomal genes\")\n",
    "\n",
    "    if ribo_genes.sum() > 0:\n",
    "        ribo_gene_names = adata.var_names[ribo_genes].tolist()\n",
    "\n",
    "        # Calculate ribosomal percentage\n",
    "        adata.obs['pct_counts_ribo'] = (adata[:, ribo_genes].X.sum(axis=1) / adata.X.sum(axis=1)) * 100\n",
    "        if hasattr(adata.obs['pct_counts_ribo'], 'A1'):\n",
    "            adata.obs['pct_counts_ribo'] = adata.obs['pct_counts_ribo'].A1\n",
    "\n",
    "        print(f\"  ‚úÖ Added ribosomal percentage and score using {len(ribo_gene_names)} genes\")\n",
    "    else:\n",
    "        raise RuntimeError(\"CRITICAL: No ribosomal genes found for scoring. Ribosomal gene metrics are required for quality control.\")\n",
    "\n",
    "    # Summary statistics\n",
    "    print(f\"  ‚úÖ Annotated {annotated_count}/{adata.n_vars} genes\")\n",
    "    print(f\"  üìä Gene types: {adata.var['gene_type'].value_counts().head().to_dict()}\")\n",
    "    print(f\"  üìä Ribosomal: {adata.var['ribosomal'].sum()}\")\n",
    "    print(f\"  üìä Cell cycle: {adata.var['cell_cycle'].sum()}\")\n",
    "    print(f\"  üìä Chromosomes: {adata.var['chromosome'].nunique()} unique\")\n",
    "    print(f\"  üìä Mean % cells expressed: {adata.var['pct_cells_expressed'].mean():.1f}%\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Matplotlib is building the font cache; this may take a moment.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded adata: AnnData object with n_obs √ó n_vars = 400356 √ó 34183\n",
      "    layers: 'ambiguous', 'mature', 'nascent'\n",
      "var index example: ['ENSMUSG00000086053.2', 'ENSMUSG00000100764.2', 'ENSMUSG00000102095.2', 'ENSMUSG00000100635.2', 'ENSMUSG00000100480.2']\n",
      "üß¨ Adding comprehensive gene annotations...\n",
      "  üìã Loaded 182 ribosomal gene\n",
      "  üìã Sample ribosomal gene: ['Mrpl41', 'Mrps28', 'Mrps10', 'Rps23', 'Rpl39l']\n",
      "  üìä Calculating % cells expressed...\n",
      "  üóÉÔ∏è  Loading gene annotations from gencode database...\n",
      "  üîç Sample gene IDs from adata.var.index: ['ENSMUSG00000086053.2', 'ENSMUSG00000100764.2', 'ENSMUSG00000102095.2', 'ENSMUSG00000100635.2', 'ENSMUSG00000100480.2']\n",
      "  üîç Trying to annotate gene 1: ENSMUSG00000086053.2\n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "Failed to annotate gene ENSMUSG00000086053.2: ENSMUSG00000086053.2",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFeatureNotFoundError\u001b[0m                      Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 80\u001b[0m, in \u001b[0;36madd_comprehensive_gene_annotations\u001b[0;34m(adata, ribosomal_genes_path, gene_database_path, cell_cycle_genes_path)\u001b[0m\n\u001b[1;32m     78\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m     79\u001b[0m     \u001b[38;5;66;03m# Query database for this gene by ID\u001b[39;00m\n\u001b[0;32m---> 80\u001b[0m     gene \u001b[38;5;241m=\u001b[39m \u001b[43mdb\u001b[49m\u001b[43m[\u001b[49m\u001b[43mgene_id\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m     81\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m i \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m5\u001b[39m:\n",
      "File \u001b[0;32m/oak/stanford/groups/engreitz/Users/shshanhe/tools/miniforge3/envs/torch-cNMF/lib/python3.9/site-packages/gffutils/interface.py:296\u001b[0m, in \u001b[0;36mFeatureDB.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    295\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m results \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 296\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m FeatureNotFoundError(key)\n\u001b[1;32m    297\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_feature_returner(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mresults)\n",
      "\u001b[0;31mFeatureNotFoundError\u001b[0m: ENSMUSG00000086053.2",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[10], line 16\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mLoaded adata:\u001b[39m\u001b[38;5;124m\"\u001b[39m, adata)\n\u001b[1;32m     15\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvar index example:\u001b[39m\u001b[38;5;124m\"\u001b[39m, adata\u001b[38;5;241m.\u001b[39mvar_names[:\u001b[38;5;241m5\u001b[39m]\u001b[38;5;241m.\u001b[39mtolist())\n\u001b[0;32m---> 16\u001b[0m \u001b[43madd_comprehensive_gene_annotations\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     17\u001b[0m \u001b[43m    \u001b[49m\u001b[43madata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     18\u001b[0m \u001b[43m    \u001b[49m\u001b[43mribosomal_genes_path\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     19\u001b[0m \u001b[43m    \u001b[49m\u001b[43mgene_database_path\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     20\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcell_cycle_genes_path\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[9], line 116\u001b[0m, in \u001b[0;36madd_comprehensive_gene_annotations\u001b[0;34m(adata, ribosomal_genes_path, gene_database_path, cell_cycle_genes_path)\u001b[0m\n\u001b[1;32m    113\u001b[0m         annotated_count \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m    115\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m--> 116\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFailed to annotate gene \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mgene_id\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00me\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01me\u001b[39;00m\n\u001b[1;32m    118\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m  üìä Annotation results: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mannotated_count\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m success, \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mmissing_count\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m missing, \u001b[39m\u001b[38;5;132;01m{\u001b[39;00merror_count\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m errors\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    120\u001b[0m \u001b[38;5;66;03m# CRITICAL: Fail if gene annotation completely failed\u001b[39;00m\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Failed to annotate gene ENSMUSG00000086053.2: ENSMUSG00000086053.2"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import gffutils\n",
    "import numpy as np\n",
    "\n",
    "gex_h5ad=\"/scratch/users/shshanhe/eRZ58_20A_analysis/eRZ58:eRZ58_GEX20A/kb_all_main_raw/counts_unfiltered_modified/adata.h5ad\"\n",
    "guide_h5ad=\"/scratch/users/shshanhe/eRZ58_20A_analysis/eRZ58:eRZ58_gRNA20A/kb_guide_main_raw/counts_unfiltered_modified/adata.h5ad\"\n",
    "gene_database_path = \"/oak/stanford/groups/engreitz/Users/shshanhe/project/exchange_perturb_project/data/reference/Mus_musculus.GRCm39.108.annotation.db\"\n",
    "db = gffutils.FeatureDB(gene_database_path)\n",
    "ribosomal_genes_path=\"/oak/stanford/groups/engreitz/Users/shshanhe/project/exchange_perturb_project/data/reference/genes_list_ribosomal.filtered.txt\"\n",
    "cell_cycle_genes_path=\"/oak/stanford/groups/engreitz/Users/shshanhe/project/exchange_perturb_project/data/reference/regev_lab_cell_cycle_genes_convert_to_mouse.txt\"\n",
    "\n",
    "\n",
    "adata = sc.read_h5ad(gex_h5ad)\n",
    "print(\"Loaded adata:\", adata)\n",
    "print(\"var index example:\", adata.var_names[:5].tolist())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß¨ Adding comprehensive gene annotations...\n",
      "  üìã Loaded 182 ribosomal gene\n",
      "  üìã Sample ribosomal gene: ['Mrpl41', 'Mrps28', 'Mrps10', 'Rps23', 'Rpl39l']\n",
      "  üìä Calculating % cells expressed...\n",
      "  üóÉÔ∏è  Loading gene annotations from gencode database...\n",
      "  üîç Sample gene IDs from adata.var.index: ['ENSMUSG00000086053.2', 'ENSMUSG00000100764.2', 'ENSMUSG00000102095.2', 'ENSMUSG00000100635.2', 'ENSMUSG00000100480.2']\n",
      "  üîç Trying to annotate gene 1: ENSMUSG00000086053.2\n",
      "    ‚úÖ Found gene ENSMUSG00000086053.2 in database\n",
      "    üìã Gene info: type=lncRNA, name=Gm15178, chr=1\n",
      "  üîç Trying to annotate gene 2: ENSMUSG00000100764.2\n",
      "    ‚úÖ Found gene ENSMUSG00000100764.2 in database\n",
      "    üìã Gene info: type=lncRNA, name=Gm29155, chr=1\n",
      "  üîç Trying to annotate gene 3: ENSMUSG00000102095.2\n",
      "    ‚úÖ Found gene ENSMUSG00000102095.2 in database\n",
      "    üìã Gene info: type=lncRNA, name=C730036E19Rik, chr=1\n",
      "  üîç Trying to annotate gene 4: ENSMUSG00000100635.2\n",
      "    ‚úÖ Found gene ENSMUSG00000100635.2 in database\n",
      "    üìã Gene info: type=lncRNA, name=Gm29157, chr=1\n",
      "  üîç Trying to annotate gene 5: ENSMUSG00000100480.2\n",
      "    ‚úÖ Found gene ENSMUSG00000100480.2 in database\n",
      "    üìã Gene info: type=lncRNA, name=Gm29156, chr=1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  üìä Annotation results: 34183 success, 0 missing, 0 errors\n",
      "  üß¨ Calculating mitochondrial gene scores...\n",
      "  üìä Found 13 mitochondrial genes\n",
      "  ‚úÖ Added MT percentage and score using 13 genes\n",
      "  üß¨ Calculating ribosomal gene scores...\n",
      "  üìä Found 176 ribosomal genes\n",
      "  ‚úÖ Added ribosomal percentage and score using 176 genes\n",
      "  ‚úÖ Annotated 34183/34183 genes\n",
      "  üìä Gene types: {'protein_coding': 21959, 'lncRNA': 11527, 'IG_V_gene': 218, 'IG_V_pseudogene': 158, 'TR_V_gene': 144}\n",
      "  üìä Ribosomal: 176\n",
      "  üìä Cell cycle: 97\n",
      "  üìä Chromosomes: 35 unique\n",
      "  üìä Mean % cells expressed: 0.0%\n"
     ]
    }
   ],
   "source": [
    "add_comprehensive_gene_annotations(\n",
    "    adata,\n",
    "    ribosomal_genes_path,\n",
    "    gene_database_path,\n",
    "    cell_cycle_genes_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File 1: /oak/stanford/groups/engreitz/Users/shshanhe/project/exchange_perturb_project/data/reference/human_ribosomal_genes_convert_to_mouse.txt\n",
      "  n = 181\n",
      "File 2: /oak/stanford/groups/engreitz/Users/shshanhe/project/exchange_perturb_project/data/reference/genes_list_ribosomal.filtered.txt\n",
      "  n = 182\n",
      "\n",
      "Overlap:\n",
      "  n = 163\n",
      "\n",
      "Examples of overlap (up to 30):\n",
      "['Fau', 'Mrpl1', 'Mrpl10', 'Mrpl11', 'Mrpl12', 'Mrpl13', 'Mrpl14', 'Mrpl15', 'Mrpl16', 'Mrpl17', 'Mrpl18', 'Mrpl19', 'Mrpl2', 'Mrpl20', 'Mrpl21', 'Mrpl22', 'Mrpl23', 'Mrpl24', 'Mrpl27', 'Mrpl28', 'Mrpl3', 'Mrpl30', 'Mrpl32', 'Mrpl33', 'Mrpl34', 'Mrpl35', 'Mrpl36', 'Mrpl37', 'Mrpl38', 'Mrpl39']\n",
      "\n",
      "Only in file 1 (up to 30):\n",
      "['Aurkaip1', 'Chchd1', 'Dap3', 'Gadd45gip1', 'Gm6133', 'Gm6525', 'Ptcd3', 'Rpl27rt', 'Rpl34-ps1', 'Rpl36-ps4', 'Rpl36a-ps1', 'Rpl37rt', 'Rpl9-ps6', 'Rplp1rt', 'Rplp2-ps1', 'Rps18-ps5', 'Rps18-ps6', 'Rps27rt']\n",
      "\n",
      "Only in file 2 (up to 30):\n",
      "['Mrps36', 'Rpl32l', 'Rpl6l', 'Rps12l1', 'Rps19bp1', 'Rps23rg1', 'Rps3a2', 'Rps3a3', 'Rps4l', 'Rps6ka1', 'Rps6ka2', 'Rps6ka3', 'Rps6ka4', 'Rps6ka5', 'Rps6ka6', 'Rps6kb1', 'Rps6kb2', 'Rps6kc1', 'Rps6kl1']\n"
     ]
    }
   ],
   "source": [
    "# Compare overlap between two ribosomal gene lists (one gene symbol per line)\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "f1 = Path(\"/oak/stanford/groups/engreitz/Users/shshanhe/project/exchange_perturb_project/data/reference/human_ribosomal_genes_convert_to_mouse.txt\")\n",
    "f2 = Path(\"/oak/stanford/groups/engreitz/Users/shshanhe/project/exchange_perturb_project/data/reference/genes_list_ribosomal.filtered.txt\")\n",
    "\n",
    "# read + clean (drop empty lines, strip whitespace, drop comments)\n",
    "s1 = {line.strip() for line in f1.read_text().splitlines() if line.strip() and not line.strip().startswith(\"#\")}\n",
    "s2 = {line.strip() for line in f2.read_text().splitlines() if line.strip() and not line.strip().startswith(\"#\")}\n",
    "\n",
    "overlap = s1 & s2\n",
    "only1 = s1 - s2\n",
    "only2 = s2 - s1\n",
    "\n",
    "print(\"File 1:\", f1)\n",
    "print(\"  n =\", len(s1))\n",
    "print(\"File 2:\", f2)\n",
    "print(\"  n =\", len(s2))\n",
    "print(\"\\nOverlap:\")\n",
    "print(\"  n =\", len(overlap))\n",
    "\n",
    "# show a few examples\n",
    "print(\"\\nExamples of overlap (up to 30):\")\n",
    "print(sorted(list(overlap))[:30])\n",
    "\n",
    "print(\"\\nOnly in file 1 (up to 30):\")\n",
    "print(sorted(list(only1))[:30])\n",
    "\n",
    "print(\"\\nOnly in file 2 (up to 30):\")\n",
    "print(sorted(list(only2))[:30])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
